src/demo/renderer.js

```diff
@@ -85,6 +85,27 @@ function drawTrainingScreen(state) {
   drawUpcomingFish(state);
 }
 
+function drawSingleFish(fishDatum, fishXPos, fishYPos, ctx) {
+  if (!fishDatum.canvas) {
+    fishDatum.canvas = document.createElement('canvas');
+    fishDatum.canvas.width = FISH_CANVAS_WIDTH;
+    fishDatum.canvas.height = FISH_CANVAS_HEIGHT;
+    loadFishImages(fishDatum.fish).then(results => {
+      const fishCtx = fishDatum.canvas.getContext('2d');
+      drawFish(
+        fishDatum.fish,
+        results,
+        fishCtx,
+        FISH_CANVAS_WIDTH / 2,
+        FISH_CANVAS_HEIGHT / 2
+      );
+      ctx.drawImage(fishDatum.canvas, fishXPos, fishYPos);
```

Comment:
For sure want to explore this and hear thoughts from people who have done this before! However, we are already using intermediate canvases to render each of the fish components so I think we'll need to rethink a lot of the renderer if we find this approach to be too problematic.

Reply:
For context, in https://github.com/code-dot-org/dance-party/pull/26 Chris identified an issue where drawing from one canvas to another caused a memory leak in Safari due to ever-accumulating "layer" data.  We didn't fully identify a workaround at the time, instead trying to do without the additional canvases.  I think we ended up using offscreen canvases for a few things in dance party, but they're treated like a resource that would never be released, so we're careful not to create them in a loop and expect them to clean up properly.

On the other hand, [not using an offscreen canvas is also expensive...](https://github.com/code-dot-org/dance-party/issues/452) It's a balancing act.  Make sure to start measuring as soon as possible.

While we're digging up Dance Party lessons, there's also [this report](https://github.com/code-dot-org/dance-party/pull/473) that `pixelDensity(1)` throws in Safari. That doesn't seem to be an issue yet but it might come up later.  Also [this problem with `tint()` on mobile Safari](https://github.com/code-dot-org/dance-party/issues/387).

Reply:
Thanks @islemaster, this is great information.  

@bethanyaconnor and @madelynkasula I think one thing we should strongly consider doing is preallocating all canvases that will use at app init time, and then never creating one in regular execution.  That way we set our high-water mark up front, and we won't be surprised by any canvas leaks later on.  Would this be feasible?

Reply:
I don't see why not. It'll be a bit complicated to maintain which canvases are "in-use" and which can be redrawn but still feasible.

Based on the wireframes it looks like we don't currently plan to have more than 25 fish on the screen at a time. We can do some tests about where that max _can_ be just so we can know whether 100 fish screen animated on the screen at the same time is feasible (I believe that answer is no).

Implementing that doesn't feel appropriate for this PR. I'm happy to work on it (or not if someone else wants to) but I don't think it should block this PR.

Reply:
^^^ bethany took the words right out of my mouth. for next steps, i see 2 options:

1. meet tomorrow or monday to discuss how we want to approach this, or
2. have one of us quickly prototype initializing all canvases up-front then managing them throughout the scene